pkg/network/dns_cache.go

```diff
@@ -192,61 +200,56 @@ func (c *reverseDNSCache) Expire(now time.Time) {
 	)
 }
 
-func (c *reverseDNSCache) getNamesForIP(ip util.Address, updatedTTL int64) []string {
+func (c *reverseDNSCache) getNamesForIP(ip util.Address) []string {
 	val, ok := c.data[ip]
 	if !ok {
 		return nil
 	}
-
-	val.expiration = updatedTTL
+	val.inUse = true
 	return val.copy()
 }
 
 type dnsCacheVal struct {
-	// opting for a []string instead of map[string]struct{} since common case is len(names) == 1
-	names      []string
-	expiration int64
+	names map[string]time.Time
+	// inUse keeps track of whether this dns cache record is currently in use by a connection.
+	// This flag is reset to false every time reverseDnsCache.Get is called.
+	// This flag is only set to true if reverseDNSCache.getNamesForIP returns this struct.
+	// If inUse is set, then this record will not be expired out.
+	inUse bool
 }
 
-func (v *dnsCacheVal) merge(name string, maxSize int) (rejected bool) {
-	normalized := strings.ToLower(name)
-	if i := sort.SearchStrings(v.names, normalized); i < len(v.names) && v.names[i] == normalized {
+func (v *dnsCacheVal) merge(name string, deadline time.Time, maxSize int) (rejected bool) {
+	if exp, ok := v.names[name]; ok {
+		if deadline.After(exp) {
+			v.names[name] = deadline
```

Comment:
Should we also set `inUse: true` here?

Reply:
Yes, I think we probably should. I'm going to do it in `Add` though to cover all cases.

Reply:
Fixed here, since I don't want to flag it in the rejected case.